### PR TITLE
Added way to disable group synchronization, to reduce extra processing

### LINKED ISSUE #657: Provide an option to disable group synchronization code paths entirely
The `resync` function takes a long amount of time to run on our main server, and we use an alternative method of synchronisation.

### PR DESCRIPTION
**Pull request by metalshark**
Added GroupRoleSynchronizationEnabled option to prevent unnecessary resync work when unused.
Fixes #657

### CODE REVIEW

File: src/main/java/github/scarsz/discordsrv/commands/CommandResync.java [hunk 1]
```diff
@@ -14,14 +14,16 @@
...
+        if (DiscordSRV.config().getBoolean("GroupRoleSynchronizationEnabled")) {
+            Bukkit.getScheduler().runTaskAsynchronously(DiscordSRV.getPlugin(), () -> {
+                sender.sendMessage(ChatColor.AQUA + "Full group synchronization triggered.");
+                long time = System.currentTimeMillis();
+                DiscordSRV.getPlugin().getGroupSynchronizationManager().resyncEveryone();
+                time = System.currentTimeMillis() - time;
+                int seconds = Math.toIntExact(TimeUnit.MILLISECONDS.toSeconds(time));
+                sender.sendMessage(ChatColor.AQUA + "Full group synchronization finished, taking " + seconds + " seconds.");
+            });
+        }
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/commands/CommandResync.java]: Should provide a reponse if disabled

File: src/main/java/github/scarsz/discordsrv/commands/CommandResync.java [hunk 2]
```diff
@@ -14,6 +14,10 @@
             permission = "discordsrv.resync"
     )
     public static void execute(CommandSender sender, String[] args) {
+        if (!DiscordSRV.config().getBoolean("GroupRoleSynchronizationEnabled")) {
+            sender.sendMessage(ChatColor.RED + "Group synchonization is disabled. Please set GroupRoleSynchronizationEnabled to true in synchronization.yml to use this feature.");
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/commands/CommandResync.java]: LangUtil.InternalMessage

**Comment by Vankka** [comment]
Commands may also make groups to resync

**Comment by Scarsz** [comment]
This should just check if there’s no groups to sync in the config list, a dedicated config option to disable isn’t necessary

File: src/main/java/github/scarsz/discordsrv/commands/CommandResync.java [hunk 3]
```diff
@@ -14,6 +15,10 @@
             permission = "discordsrv.resync"
     )
     public static void execute(CommandSender sender, String[] args) {
+        if (DiscordSRV.config().getMap("GroupRoleSynchronizationGroupsAndRolesToSync").isEmpty()) {
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/commands/CommandResync.java]: Would be better to also check that the ids aren't/are valid incase the option is on default values

Comment (metalshark) [reply src/main/java/github/scarsz/discordsrv/commands/CommandResync.java]: Included a check on default and a check for roleId. Feel free to pick which you'd prefer (as both seems unnecessary).

File: src/main/java/github/scarsz/discordsrv/DiscordSRV.java [hunk 4]
```diff
@@ -1077,4 +1077,26 @@ public static boolean isUpdateCheckDisabled() {
...
+     */
+    public static boolean isGroupRoleSynchronizationEnabled() {
+        final Map<String, String> groupsAndRolesToSync = DiscordSRV.config().getMap("GroupRoleSynchronizationGroupsAndRolesToSync");
+        if (groupsAndRolesToSync.isEmpty()) return false;
+        if (groupsAndRolesToSync.equals(DEFAULT_GROUP_AND_ROLES_TO_SYNC)) return false;
+        for(Map.Entry<String, String> entry : groupsAndRolesToSync.entrySet()) {
+            final String group = entry.getKey();
+            final String roleId = entry.getValue();
+            if (group.isEmpty()) return false;
+            if (roleId.isEmpty() || roleId.equals(DEFAULT_ROLE_ID)) return false;
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/DiscordSRV.java]: This returns false even when there is a valid id and one or more default values

Comment (metalshark) [reply src/main/java/github/scarsz/discordsrv/DiscordSRV.java]: We have choices:
1. only look for default configuration and/or completely blank
2. look for any role being 18 x 0's
3. look for at least one role not being 18 x 0's
am guessing you'd prefer option 3?

Comment (Vankka) [reply src/main/java/github/scarsz/discordsrv/DiscordSRV.java]: number 3 would be ideal here, yes

Comment (metalshark) [reply src/main/java/github/scarsz/discordsrv/DiscordSRV.java]: How do you feel about the constant `DEFAULT_ROLE_ID` being declared? Can easily get rid of it and just put the string in instead.

Comment (Vankka) [reply src/main/java/github/scarsz/discordsrv/DiscordSRV.java]: doesn't really make a difference, though using a constant is inconsistent with other checks in DiscordSRV

**Comment by metalshark** [comment]
Thinking about this again, I should probably cache the result of the `isGroupRoleSynchronizationEnable()` check and refresh it when the config is reloaded.

**Comment by Vankka** [comment]
Would be inconsistent with other implementations, and it's a fairly minor speed bump

**Pull request merged**